riscv_executor/src/lib.rs

```diff
@@ -105,36 +105,83 @@ pub enum MemOperationKind {
 
 #[derive(Debug)]
 pub struct MemOperation {
-    /// Line of the register trace the memory operation happened.
-    pub idx: usize,
+    /// The row of the execution trace the memory operation happened.
+    pub row: usize,
     pub kind: MemOperationKind,
     pub address: u32,
 }
 
+pub struct RegWrite {
+    /// The row of the execution trace this write will effect. Multiple writes
+    /// at the same row are valid: the last write to a given reg_idx will define
+    /// the final value of the register in that row.
+    row: usize,
+    /// Index of the register in the register bank written.
+    reg_idx: u16,
+    val: Elem,
+}
+
 pub struct ExecutionTrace {
-    pub reg_map: HashMap<String, usize>,
+    pub reg_map: HashMap<String, u16>,
 
     /// Values of the registers in the execution trace.
     ///
     /// Each N elements is a row with all registers, where N is the number of
     /// registers.
-    pub regs: Vec<Elem>,
+    pub reg_writes: Vec<RegWrite>,
 
     /// Writes and reads to memory.
-    pub mem: Vec<MemOperation>,
+    pub mem_ops: Vec<MemOperation>,
 
-    /// The length of the trace.
+    /// The length of the trace, after applying the reg_writes.
     pub len: usize,
 }
 
 impl ExecutionTrace {
-    /// Split the values of the registers' trace into rows.
-    pub fn regs_rows(&self) -> impl Iterator<Item = &[Elem]> {
-        self.regs.chunks_exact(self.reg_map.len())
+    /// Replay the execution and get the register values per trace row.
+    pub fn replay(&self) -> TraceReplay {
+        TraceReplay {
+            trace: self,
+            regs: vec![0.into(); self.reg_map.len()],
```

Comment:
Why 0? Wasn't this supposed to replay something?

Reply:
Zero is the initial state of all registers. This vector only holds one row at a time.

Reply:
what is this function supposed to do? Just create the object?

Reply:
Yes, this object was supposed to be an iterator over the trace rows, but I couldn't implement `Iterator` on it because `next()` would have to return a slice borrowed from `self`, and the iterator interface doesn't allow for that, as the result of a proper `Iterator` should be able to outlive the iterator itself.

So now it is a plain struct with a method called `next_row()` instead of `next()`, because clippy complains that it might be confused with `next()` from trait `Iterator` (and suggest me to implement `Iterator` for it instead, which is impossible).

Reply:
if this function only creates the object, why not call it `new`?

Reply:
Because it creates an object of another type, not `Self`.

The other object is an iterator-like, and the convention is the thing they are iterating on to create them. E.g., it is `vec.drain()`, not `Drain::new(&mut vec)`.

Reply:
ok